arkouda/timeclass.py

```diff
@@ -144,6 +143,158 @@ def __init__(self, array, unit: str = _BASE_UNIT):  # type: ignore
+            self._second %= 60
+
+            # Month, day
+            self._day = self.values // _get_factor("d")
+            self._day += 1401  # move starting point to 1996 - 03 - 01
+            self._day %= 1461  # max 3 years 365 day (not 4 years!) remaining
+            d = self._day[:]
+            self._day[d >= 731] -= 731
+
+            d = self._day[:]
+            self._day[d >= 365] -= 365
+
+            d = self._day[:]
+            self._month = d[:] - d[:] + 3
+            self._day[d >= 306] -= 306
+            self._month[d >= 306] = 1
+
+            d = self._day[:]
+            self._day[d >= 153] -= 153
+            self._month[d >= 153] = 8
+
+            d = self._day[:]
+            self._day[d >= 122] -= 122
+            self._month[d >= 122] += 4
+            d = self._day[:]
+            self._day[d >= 61] -= 61
+            self._month[d >= 61] += 2
+
+            d = self._day[:] + 2
+            self._month[d >= 32] += 1
+            months = self._month[:]
+            months[months >= 13] -= 1
+            self._month[self._month >= 13] = 1
+            self._day += 2
+            self._day[d >= 32] -= 31
+
+            # Day of year
+            month_days = [
+                [0, 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334],
+                [0, 0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335],
+            ]
+
+            lyear1 = self._year % 4 == 0
+            lyear2 = self._year % 100 != 0
+            lyear3 = self._year % 400 == 0
+            lyear = lyear1 & lyear2
+            lyear = cast((lyear | lyear3), int64)
+
+            days = self._day[:]
+            self._date = []
+            for i in range(days.size):
+                if self._month[i] == 2 and self._day[i] == 29 and lyear[i] == 0:
+                    self._month[i] += 1
+                    self._day[i] = 1
+                    days[i] += month_days[lyear[i]][self._month[i] - 1]
+                else:
+                    days[i] += month_days[lyear[i]][self._month[i]]
+                self._date.append(f"{self._year[i]}-{self._month[i]:02d}-{self._day[i]:02d}")
+            self._dayofyear = ak.array(days)
+
+            # Day of week
+            self._dayofweek = cast(((self.values / _get_factor("d")) + 3), int64)
+            self._dayofweek %= 7
+
+            # Week of year
+            week = (self._dayofyear[:] - 1) - self._dayofweek + 3
+            week[week >= 0] //= 7
+            week[week >= 0] += 1
+
+            for i in range(days.size):
+                if week[i] < 0:
+                    year = int(self._year[i] - 1)
+                    if week[i] > -2 or (week[i] == -2 and _isLeapYear(year)):
+                        week[i] = 53
+                    else:
+                        week[i] = 52
+                elif week[i] == 53:
+                    if 31 - self._day[i] + self.dayofweek[i] < 3:
+                        week[i] = 1
```

Comment:
This is something we are definitely going to need to move to the server. @pierce314159 do you have any strong opinions on if it goes no or in a latter PR?

Reply:
yeah i think we should def move this and the other `for i in range(days.size)` loop to the server. I would be okay doing that in another PR, but I think this is still not performant enough to go in rn. My thinking is if we can come up with a way to do these loops with vector operations, then it should be fine to keep them client side for now. But if we can't come up with a vector op way, we should move it to the server in this PR